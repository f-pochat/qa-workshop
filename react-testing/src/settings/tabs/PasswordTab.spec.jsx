import React from 'react'
import { render, screen, fireEvent, waitFor, renderHook } from '@/test'
import { describe, expect, test, vi } from 'vitest'
import userEvent from '@testing-library/user-event'
import { PasswordTab } from '.';
import { useToast } from '@/components/ui/use-toast';

const mockToast = vi.fn()
vi.mock('@/components/ui/use-toast', () => ({
    useToast: vi.fn(() => ({
        toast: mockToast
    })),
}));



describe('Passwords Card', () => {
    test('should match snapshot', () => {
        const { container } = render(<PasswordTab />)
        expect(container.firstChild).toMatchSnapshot();
    })

    test('Submitting passwords, toast is called', async () => {
        const { getByPlaceholderText, getByRole } = render(<PasswordTab />);

        await userEvent.type(getByPlaceholderText('Current Password'), 'nicou')
        await userEvent.type(getByPlaceholderText('New Password'), 'Nicou1234!')
        await userEvent.type(getByPlaceholderText('Repeat Password'), 'Nicou1234!')

        fireEvent.click(getByRole('button', { name: "Submit" }))

        await waitFor(() => {
            expect(mockToast).toHaveBeenCalledWith({
                title: "Success",
                text: "Password changed"
            });
        });
    });

    test('Submitting same current and new password', async () => {
        const { getByText, getByRole, getByPlaceholderText } = render(<PasswordTab />);

        await userEvent.type(getByPlaceholderText('Current Password'), 'nicou')
        await userEvent.type(getByPlaceholderText('New Password'), 'nicou')
        await userEvent.type(getByPlaceholderText('Repeat Password'), 'Nicou1234!')

        fireEvent.click(getByRole('button', { name: "Submit" }))

        await waitFor(() => {
            expect(getByText("New password must be different from current password")).toBeInTheDocument()
            expect(getByText("Passwords must match")).toBeInTheDocument()
        })
    });
})