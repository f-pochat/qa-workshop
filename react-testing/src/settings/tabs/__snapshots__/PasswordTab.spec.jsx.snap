// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Passwords Card > should match snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
    >
      Password
    </h3>
    <p
      class="text-sm text-muted-foreground"
    >
      Change your password here. After saving, you'll be logged out.
    </p>
  </div>
  <div
    class="p-6 pt-0 space-y-2"
  >
    <form
      class="space-y-8"
    >
      <div
        class="space-y-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for=":r0:-form-item"
        >
          Current Password
        </label>
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r0:-form-item"
          name="currentPassword"
          placeholder="Current Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for=":r1:-form-item"
        >
          New Password
        </label>
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r1:-form-item"
          name="newPassword"
          placeholder="New Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for=":r2:-form-item"
        >
          Repeat Password
        </label>
        <input
          aria-describedby=":r2:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r2:-form-item"
          name="repeatPassword"
          placeholder="Repeat Password"
          type="password"
          value=""
        />
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
